base/reducedim.jl

```diff
@@ -125,46 +125,38 @@ function _reducedim_init(f, op, fv, fop, A, region)
 end
 
 # initialization when computing minima and maxima requires a little care
-for (f1, f2, initval, typeextreme) in ((:min, :max, :Inf, :typemax), (:max, :min, :(-Inf), :typemin))
-    @eval function reducedim_init(f, op::typeof($f1), A::AbstractArray, region)
-        # First compute the reduce indices. This will throw an ArgumentError
-        # if any region is invalid
-        ri = reduced_indices(A, region)
+function reducedim_init(f::F, ::Union{typeof(min),typeof(max)}, A::AbstractArray, region) where {F}
+    # First compute the reduce indices. This will throw an ArgumentError
+    # if any region is invalid
+    ri = reduced_indices(A, region)
 
-        # Next, throw if reduction is over a region with length zero
-        any(i -> isempty(axes(A, i)), region) && _empty_reduce_error()
+    # Next, throw if reduction is over a region with length zero
+    any(i -> isempty(axes(A, i)), region) && _empty_reduce_error()
 
-        # Make a view of the first slice of the region
-        A1 = view(A, ri...)
+    # Make a view of the first slice of the region
+    A1 = view(A, ri...)
 
-        if isempty(A1)
-            # If the slice is empty just return non-view version as the initial array
-            return map(f, A1)
-        else
-            # otherwise use the min/max of the first slice as initial value
-            v0 = mapreduce(f, $f2, A1)
-
-            T = _realtype(f, promote_union(eltype(A)))
-            Tr = v0 isa T ? T : typeof(v0)
-
-            # but NaNs and missing need to be avoided as initial values
-            if v0 isa Number && isnan(v0)
-                # v0 is NaN
-                v0 = oftype(v0, $initval)
-            elseif isunordered(v0)
-                # v0 is missing or a third-party unordered value
-                Tnm = nonmissingtype(Tr)
-                # TODO: Some types, like BigInt, don't support typemin/typemax.
-                # So a Matrix{Union{BigInt, Missing}} can still error here.
-                v0 = $typeextreme(Tnm)
-            end
-            # v0 may have changed type.
-            Tr = v0 isa T ? T : typeof(v0)
+    # calculate the output type
+    T = promote_typejoin_union(_return_type(f, Tuple{eltype(A)}))
```

Comment:
Please don't use `_return_type` in a way that the user can observe the precision of the inference.

Reply:
Ummm, but we need it to know the `eltype` of output array.
Is there a better way to do this?

Reply:
In fact the following code cause crash on 1.7:
```julia
julia> a = collect(1.0:4);

julia> f(x) = x < 3 ? missing : x;

julia> minimum(f, a; dims = 1)
<unknown>:0: error: starting new .cfi frame before finishing the previous one
<unknown>:0: error: this directive must appear between .cfi_startproc and .cfi_endproc directives
<unknown>:0: error: starting new .cfi frame before finishing the previous one
<unknown>:0: error: this directive must appear between .cfi_startproc and .cfi_endproc directives

Please submit a bug report with steps to reproduce this fault, and any error messages that follow (in their entirety). Thanks.
Exception: EXCEPTION_ACCESS_VIOLATION at 0x0 -- unknown function (ip: 0000000000000000)
in expression starting at none:0
unknown function (ip: 0000000000000000)
warning: DWARF compile unit extends beyond its bounds cu 0x00000000 at 0x0000000b

error: No unit DIE
unknown function (ip: 000000004eaf2fb7)
#current_exceptions#67 at .\error.jl:155
current_exceptions at .\error.jl:149 [inlined]
current_exceptions at .\error.jl:149 [inlined]
eval_user_input at C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.7\REPL\src\REPL.jl:162
repl_backend_loop at C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.7\REPL\src\REPL.jl:244
start_repl_backend at C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.7\REPL\src\REPL.jl:229
#run_repl#47 at C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.7\REPL\src\REPL.jl:362
run_repl at C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.7\REPL\src\REPL.jl:349
#930 at .\client.jl:394
jfptr_YY.930_35191.clone_1 at C:\Users\MYJ\AppData\Local\Programs\Julia-1.7.1\lib\julia\sys.dll (unknown line)
jl_apply at /cygdrive/c/buildbot/worker/package_win64/build/src\julia.h:1788 [inlined]
jl_f__call_latest at /cygdrive/c/buildbot/worker/package_win64/build/src\builtins.c:757
#invokelatest#2 at .\essentials.jl:716 [inlined]
invokelatest at .\essentials.jl:714 [inlined]
run_main_repl at .\client.jl:379
exec_options at .\client.jl:309
_start at .\client.jl:495
jfptr__start_43221.clone_1 at C:\Users\MYJ\AppData\Local\Programs\Julia-1.7.1\lib\julia\sys.dll (unknown line)
jl_apply at /cygdrive/c/buildbot/worker/package_win64/build/src\julia.h:1788 [inlined]
true_main at /cygdrive/c/buildbot/worker/package_win64/build/src\jlapi.c:559
jl_repl_entrypoint at /cygdrive/c/buildbot/worker/package_win64/build/src\jlapi.c:701
mainCRTStartup at /cygdrive/c/buildbot/worker/package_win64/build/cli\loader_exe.c:42
BaseThreadInitThunk at C:\WINDOWS\System32\KERNEL32.DLL (unknown line)
RtlUserThreadStart at C:\WINDOWS\SYSTEM32\ntdll.dll (unknown line)
Allocations: 2401717 (Pool: 2400926; Big: 791); GC: 3
```

Reply:
I guess #35323 is able to fix the crash.
But it still throw error unneeded for the following
```julia
julia> a = range(.0,1,4);

julia> f(x) = x < 0.5 ? 1 : x;

julia> minimum(f, a; dims = 1)
ERROR: InexactError: Int64(0.6666666666666666)
```
If `f` is unstable, I guess `Array{Real}` or `Array{Any}` is better than a `ERROR`?

Reply:
I try to look into the histroy of preallocation in `Base`.
Looks like inference based initialization is prefered only for empty cases.
But for `mapreduce`, if we alow user to change the elements type of inputs (`sum(Float64, [1,2,3]; dims = 1)`).
I guess we'd better not only use "the first slice" to determine the output eltype, as any unstable `f` might cause problem.
For example:
```julia
julia> sum(x -> x < 0.5 ? 0 : x, range(0.0,1,4), dims = 1)
ERROR: InexactError: Int64(1.6666666666666665)
```
User (like me) might argue that `Vector{Real}` or `Vector{Float64}` is a better choise for preallocation.
I saw #29739 removes
 > !!! warning
    In pathological cases, the type returned by `promote_op(f, argtypes...)` may not even
    be a supertype of the return value of `f(::argtypes...)`. Therefore, `promote_op`
    should _not_ be used e.g. in the preallocation of an output array.

@vtjnash is our inference system more reliable today?

Reply:
@tkf. Inference based preallocation is limited to concrete cases. Otherwise use the current logic to derive eltype.
I'm not sure if this is better.

Reply:
If you check if `_return_type` returns a concrete type, you are leaking the implementation detail of the compiler that the given function can be inferred to return a concrete type. It is true that Base and stdlib have a lot of usages of the compiler API that leak what the compiler does. It may be even possible to argue that this is a necessary evil and removing such usages would have delayed Julia 1.0 release by years. But I just don't want to contribute to increasing invalid usages myself (at least until all other options are explored).

> @vtjnash is our inference system more reliable today?

This question is irrelevant in this context. Correct Julia code does not rely on what the compiler does.

Reply:
Well then I'll use the current code for initialization, and mark some test as broken.